src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -25,37 +28,174 @@ object PV63 {
+      s"""Account{
+         |nonce: $nonce
+         |balance: $balance wei
+         |storageRoot: ${Hex.toHexString(storageRoot.toArray[Byte])}
+         |codeHash: ${Hex.toHexString(codeHash.toArray[Byte])}
          |}
        """.stripMargin
     }
   }
 
+  object MptNode {
+    implicit val rlpEndDec = new RLPEncoder[MptNode] with RLPDecoder[MptNode] {
+      override def encode(obj: MptNode): RLPEncodeable = {
+        obj match {
+          case n: MptLeaf =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = true)), value.fold(a => rlpEncode[Account](a), b => b: RLPEncodeable))
+          case n: MptExtension =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), child.fold(_.hash, _.value): RLPEncodeable)
+          case n: MptBranch =>
+            import n._
+            RLPList(children.map(e => e.fold(_.hash, _.value)).map(e => byteStringEncDec.encode(e)) :+ (terminator: RLPEncodeable): _*)
+        }
+      }
+
+      override def decode(rlp: RLPEncodeable): MptNode = rlp match {
+        case rlpList: RLPList if rlpList.items.length == 17 =>
+          MptBranch(rlpList.items.take(16).map { bytes =>
+            val v = rlpDecode[ByteString](bytes)
+            if (v.nonEmpty && v.length < 32) {
+              Right(MptValue(v))
+            } else {
+              Left(MptHash(v))
+            }
+          }, byteStringEncDec.decode(rlpList.items(16)))
+        case RLPList(hpEncoded, value) =>
+          hpDecode(hpEncoded: Array[Byte]) match {
+            case (decoded, true) =>
+              MptLeaf(ByteString(decoded),
+                Try(Left(rlpDecode[Account](value: Array[Byte])))
+                  .getOrElse(Right(rlpDecode[ByteString](value))))
+            case (decoded, false) =>
+              val v = rlpDecode[ByteString](value)
+              val child = if (v.nonEmpty && v.length < 32)
+                Right(MptValue(v))
+              else
+                Left(MptHash(v))
+
+              MptExtension(ByteString(decoded), child)
+          }
+        case _ =>
+          throw new RuntimeException("Cannot decode NodeData")
+      }
+    }
+  }
+
+  trait MptNode
+
   object NodeData {
     implicit val rlpEndDec = new RLPEncoder[NodeData] with RLPDecoder[NodeData] {
       override def encode(obj: NodeData): RLPEncodeable = {
         import obj._
-        RLPList(values: _*)
+
+        RLPList(values.map {
+          case Left(node) => RLPValue(rlpEncode[MptNode](node))
+          case Right(byteValue) => RLPValue(byteValue.toArray[Byte])
+        }: _*)
       }
 
       override def decode(rlp: RLPEncodeable): NodeData = rlp match {
-        case rlpList: RLPList => NodeData(rlpList.items)
+        case rlpList: RLPList =>
+          NodeData(rlpList.items.map { e =>
+            Try {
+              val v = rawDecode(e: Array[Byte])
+              Left(MptNode.rlpEndDec.decode(v))
```

Comment:
Minor detail: Would it be the same to use `rlpDecode[MptNode](e: Array[Byte])` instead of `MptNode.rlpEndDec.decode(v)` and `rawDecode(e: Array[Byte])`? That way we could keep `rawDecode` as an internal function for now.